<html>
<head>
<title>RNA-Seq gene expression estimation with read 
  mapping uncertainty</title>

<!-- Google Analytics -->
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-68798040-1', 'auto');
  ga('send', 'pageview');
</script>
<!-- End Google Analytics -->

</head>
<body>
<h2>RSEM (RNA-Seq by Expectation-Maximization)</h2>
<hr>
<p><strong>Updates</strong></p>
<p>Oct 3, 2016 &nbsp; RSEM v1.3.0 is online now. Added Prior-Enhanced RSEM (pRSEM) as a submodule. Introduced `--strandedness &lt;none|forward|reverse&gt;` option, `--strand-specific` and `--forward-prob` are deprecated (but still supported). Revised documentation for `rsem-plot-model`, maked it clear that in alignment statistics, isoform-level (instead of genome-level) multi-mapping reads are shown. Significantly improved the output information of `rsem-sam-validator`: if indels/clippings/skips are detected in alignments or alignments exceed transcript boundaries, `rsem-sam-validator` will report them instead of telling you the input is valid. Updated the warning message to ask users to make sure that they align their reads agains a set of transcripts instead of genome when RSEM finds less sequences in the BAM file than RSEM's indices.</p>
<p>Jun 4, 2016 &nbsp; RSEM v1.2.31 is online now. Rewrote 'rsem-gff3-to-gtf' to handle a more general set of GFF3 files. Added safety checks to make sure poly(A) tails are not added to the reference when '--star' is set.</p>
<p>Apr 24, 2016 &nbsp; Notification: Since RSEM v1.2.27, '--sort-bam-by-coordinate' needs to be set if sorted BAMs are required.</p>
<p>Apr 20, 2016 &nbsp; RSEM v1.2.30 is online now. Fixed a bug that can cause SAMtools sort to fail. Improved the appearance of warning messages: for the same type of warning messages, only show the first 50 messages and then provide the total number of such warnings.</p>
<p>Click <a href="updates.html">here</a> for full update information. </p>
<p><strong>Source Code</strong></p>
<ul>
<li><a href="https://github.com/deweylab/RSEM/archive/v1.3.0.tar.gz" onclick="ga('send', 'event', { eventCategory: 'download', eventAction: 'click', eventLabel: 'latest source codes', nonIteraction: true });">Latest version</a>
<li><a href="https://github.com/deweylab/RSEM/archive/v1.1.17.tar.gz">Version used by RSEM Galaxy Wrapper (RSEM v1.1.17)</a>
<li><a href="https://github.com/deweylab/RSEM/releases">Archive of older versions</a>
<li><a href="https://github.com/deweylab/RSEM">RSEM GitHub
    repository<a/>
</ul>
<p><strong>Documentation</strong></p>
<ul>
  <li><a href="README.html">README</a>
  <li><a href="https://github.com/bli25ucb/RSEM_tutorial">Tutorial</a>
</ul>
<p><strong>Authors</strong></p>
<p><a href="http://bli25ucb.github.io/">Bo Li</a> and <a href="https://www.biostat.wisc.edu/~cdewey/">Colin Dewey</a> designed the RSEM algorithm. <a href="http://bli25ucb.github.io/">Bo Li</a> implemented the RSEM software. <a href="https://www.biostat.wisc.edu/~cdewey/group.html">Peng Liu</a> contributed the STAR aligner options.</p>
<p><strong>License</strong></p>
<p>RSEM is under the <a href="http://www.gnu.org/copyleft/gpl.html">GNU General Public License</a></p>
<p><strong>Prebuilt RSEM Indices (RSEM v1.1.17) for Galaxy Wrapper</strong></p>
<p>These indices are based on RefSeq containing NM accession numbers only.
That means only curated genes (no experimental, no miRNA, no noncoding).
Only mature RNAs. In addition, 125bp Poly(A) tails are added at the end of each transcript.</p>
<p><a href="http://deweylab.biostat.wisc.edu/rsem/mouse_refseq_NMonly_125bpPolyATail_extractedFromMouseGenome_mm9.tar.gz">Mouse Indices</a>, extracted from mouse genome mm9</p>
<p><a href="http://deweylab.biostat.wisc.edu/rsem/human_refseq_NMonly_125bpPolyATail_extractedFromHumanGenome_hg18.tar.gz">Human Indices</a>, extracted from human genome hg18</p>
<p><strong>Reference annotations and Simulation Data used in the paper</strong></p>
<p><a href="http://deweylab.biostat.wisc.edu/rsem/reference_gtfs.tar.gz">RefSeq and Ensembl annotation GTF files used in the paper</a></p>
<p><a href="http://deweylab.biostat.wisc.edu/rsem/simulation_data.tar.gz">Simulation Data using Refseq set as reference</a></p>
<p><a href="http://deweylab.biostat.wisc.edu/rsem/simulation_data_ensembl.tar.gz">Simulation Data using Ensembl set as reference</a></p>
<p><strong>Google Users and Announce Groups</strong></p>
<table border=0 style="background-color: #fff; padding: 5px;" cellspacing=0>
  <tr><td>
  <img src="groups_logo_sm.gif"
         height=30 width=140 alt="Google Groups">
  </td></tr>
  <tr><td style="padding-left: 5px">
  <b>Subscribe to RSEM Announce</b>
  </td></tr>
  <form action="http://groups.google.com/group/rsem-announce/boxsubscribe">
  <tr><td style="padding-left: 5px;">
  Email: <input type=text name=email>
  <input type=submit name="sub" value="Subscribe">
  </td></tr>
</form>
<tr><td align=right>
  <a href="http://groups.google.com/group/rsem-announce">Visit this group</a>
</td></tr>
</table>
<table border=0 style="background-color: #fff; padding: 5px;" cellspacing=0>
  <tr><td>
  <img src="groups_logo_sm.gif"
         height=30 width=140 alt="Google Groups">
  </td></tr>
  <tr><td style="padding-left: 5px">
  <b>Subscribe to RSEM Users</b>
  </td></tr>
  <form action="http://groups.google.com/group/rsem-users/boxsubscribe">
  <tr><td style="padding-left: 5px;">
  Email: <input type=text name=email>
  <input type=submit name="sub" value="Subscribe">
  </td></tr>
</form>
<tr><td align=right>
  <a href="http://groups.google.com/group/rsem-users">Visit this group</a>
</td></tr>
</table>
<hr>
<p>(last modified on Oct 3, 2016)</p>
</body>
</html>
