<html>
<head>
<title>RNA-Seq gene expression estimation with read 
  mapping uncertainty</title>

<!-- Google Analytics -->
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-68798040-1', 'auto');
  ga('send', 'pageview');
</script>
<!-- End Google Analytics -->

</head>
<body>
<h2>RSEM (RNA-Seq by Expectation-Maximization)</h2>
<hr>
<p><strong>Updates</strong></p>
<p>Feb 1, 2016 &nbsp; RSEM v1.2.27 is online now. Upgraded SAMtools to v1.3; RSEM now supports input alignments in SAM/BAM/CRAM format. '--sam/--bam' options of 'rsem-calculate-expression' are obsoleted; use '--alignments' instead; '--sam/--bam' can still be used for compatibility with previous versions. Some 'rsem-calculate-expression' options are renamed for better interpretability. Documents are updated to reflect the SAMtools upgrade. Fixed a bug for parsing GTF files. Fixed a bug for generating transcript wiggle plots.</p>
<p>Jan 7, 2016 &nbsp; RSEM v1.2.26 is online now. RSEM supports GFF3 annotation format now. Added instructions to build RSEM references using RefSeq, Ensembl, or GENCODE annotations. Added an option to extract only transcripts from certain resources given a GTF/GFF3 file. Fixed a bug and improved the clarity of error messages for extracting transcripts from genome.</p>
<p>Nov 11, 2015 &nbsp; RSEM v1.2.25 is online now. RSEM will extract gene_name/transcript_name from GTF file when possible; however, it only appends them to the 'sample_name.*.results' files if '--append-names' option is specified; unlike v1.2.24, this version is compatible with STAR aligner even when '--append-names' is set.</p>
<p>Click <a href="updates.html">here</a> for full update information. </p>
<p><strong>Source Code</strong></p>
<ul>
<li><a href="https://github.com/deweylab/RSEM/archive/v1.2.27.tar.gz" onclick="ga('send', 'event', { eventCategory: 'download', eventAction: 'click', eventLabel: 'latest source codes', nonIteraction: true });">Latest version</a>
<li><a href="https://github.com/deweylab/RSEM/archive/v1.1.17.tar.gz">Version used by RSEM Galaxy Wrapper (RSEM v1.1.17)</a>
<li><a href="https://github.com/deweylab/RSEM/releases">Archive of older versions</a>
<li><a href="https://github.com/deweylab/RSEM">RSEM GitHub
    repository<a/>
</ul>
<p><strong>Documentation</strong></p>
<ul>
  <li><a href="README.html">README</a>
  <li><a href="https://github.com/bli25ucb/RSEM_tutorial">Tutorial</a>
</ul>
<p><strong>Authors</strong></p>
<p><a href="http://bli25ucb.github.io/">Bo Li</a> and <a href="https://www.biostat.wisc.edu/~cdewey/">Colin Dewey</a> designed the RSEM algorithm. <a href="http://bli25ucb.github.io/">Bo Li</a> implemented the RSEM software. <a href="https://www.biostat.wisc.edu/~cdewey/group.html">Peng Liu</a> contributed the STAR aligner options.</p>
<p><strong>License</strong></p>
<p>RSEM is under the <a href="http://www.gnu.org/copyleft/gpl.html">GNU General Public License</a></p>
<p><strong>Prebuilt RSEM Indices (RSEM v1.1.17) for Galaxy Wrapper</strong></p>
<p>These indices are based on RefSeq containing NM accession numbers only.
That means only curated genes (no experimental, no miRNA, no noncoding).
Only mature RNAs. In addition, 125bp Poly(A) tails are added at the end of each transcript.</p>
<p><a href="http://deweylab.biostat.wisc.edu/rsem/mouse_refseq_NMonly_125bpPolyATail_extractedFromMouseGenome_mm9.tar.gz">Mouse Indices</a>, extracted from mouse genome mm9</p>
<p><a href="http://deweylab.biostat.wisc.edu/rsem/human_refseq_NMonly_125bpPolyATail_extractedFromHumanGenome_hg18.tar.gz">Human Indices</a>, extracted from human genome hg18</p>
<p><strong>Reference annotations and Simulation Data used in the paper</strong></p>
<p><a href="http://deweylab.biostat.wisc.edu/rsem/reference_gtfs.tar.gz">RefSeq and Ensembl annotation GTF files used in the paper</a></p>
<p><a href="http://deweylab.biostat.wisc.edu/rsem/simulation_data.tar.gz">Simulation Data using Refseq set as reference</a></p>
<p><a href="http://deweylab.biostat.wisc.edu/rsem/simulation_data_ensembl.tar.gz">Simulation Data using Ensembl set as reference</a></p>
<p><strong>Google Users and Announce Groups</strong></p>
<table border=0 style="background-color: #fff; padding: 5px;" cellspacing=0>
  <tr><td>
  <img src="groups_logo_sm.gif"
         height=30 width=140 alt="Google Groups">
  </td></tr>
  <tr><td style="padding-left: 5px">
  <b>Subscribe to RSEM Announce</b>
  </td></tr>
  <form action="http://groups.google.com/group/rsem-announce/boxsubscribe">
  <tr><td style="padding-left: 5px;">
  Email: <input type=text name=email>
  <input type=submit name="sub" value="Subscribe">
  </td></tr>
</form>
<tr><td align=right>
  <a href="http://groups.google.com/group/rsem-announce">Visit this group</a>
</td></tr>
</table>
<table border=0 style="background-color: #fff; padding: 5px;" cellspacing=0>
  <tr><td>
  <img src="groups_logo_sm.gif"
         height=30 width=140 alt="Google Groups">
  </td></tr>
  <tr><td style="padding-left: 5px">
  <b>Subscribe to RSEM Users</b>
  </td></tr>
  <form action="http://groups.google.com/group/rsem-users/boxsubscribe">
  <tr><td style="padding-left: 5px;">
  Email: <input type=text name=email>
  <input type=submit name="sub" value="Subscribe">
  </td></tr>
</form>
<tr><td align=right>
  <a href="http://groups.google.com/group/rsem-users">Visit this group</a>
</td></tr>
</table>
<hr>
<p>(last modified on Feb 1, 2016)</p>
</body>
</html>
