<html>
<head>
<title>RNA-Seq gene expression estimation with read 
  mapping uncertainty</title>

<!-- Google Analytics -->
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-68798040-1', 'auto');
  ga('send', 'pageview');
</script>
<!-- End Google Analytics -->

</head>
<body>
<h2>RSEM (RNA-Seq by Expectation-Maximization)</h2>
<hr>
<p><strong>Updates</strong></p>
<p>Nov 6, 2015 &nbsp; RSEM v1.2.24 is online now. RSEM will extract gene_name/transcript_name from GTF file when possible; if extracted, gene_name/transcript_name will append at the end of gene_id/transcript_id with an underscore in between. Modified 'rsem-plot-model' to indicate the modes of fragment length and read length distributions. Modified 'rsem-plot-model' to present alignment statistics better using both barplot and pie chart. Updated 'EBSeq' to version 1.2.0. Added coefficient of quartile variation in addition to credibility intervals when '--calc-ci' is turned on. Added '--single-cell-prior' option to notify RSEM to use a sparse prior (Dir(0.1)) for single cell data; this option only makes sense if '--calc-pme' or '--calc-ci' is set.</p>
<p>Oct 19, 2015 &nbsp; RSEM v1.2.23 is online now. Moved version information from WHAT_IS_NEW to rsem_perl_utils.pm in order to make sure the '--version' option always output the version information. Fixed a typo in 'rsem-calculate-expression' that can lead an error when '--star' is set and '--star-path' is not set. Fixed a bug that can occasionally crash the RSEM simulator. Added user-friendly error messages that are triggered when RSEM detects invalid bases in the input FASTA file during reference building.</p>
<p>Jul 27, 2015 &nbsp; RSEM v1.2.22 is online now. Added options to run the STAR aligner.</p>
<p>Click <a href="updates.html">here</a> for full update information. </p>
<p><strong>Authors</strong></p>
<p><a href="http://bli25ucb.github.io/">Bo Li</a> and <a href="https://www.biostat.wisc.edu/~cdewey/">Colin Dewey</a> designed the RSEM algorithm. <a href="http://bli25ucb.github.io/">Bo Li</a> implemented the RSEM software. <a href="https://www.biostat.wisc.edu/~cdewey/group.html">Peng Liu</a> contributed the STAR aligner options.</p>
<p><strong>License</strong></p>
<p>RSEM is under the <a href="http://www.gnu.org/copyleft/gpl.html">GNU General Public License</a></p>
<p><strong>Source Code</strong></p>
<ul>
<li><a href="https://github.com/deweylab/RSEM/archive/v1.2.24.tar.gz" onclick="ga('send', 'event', { eventCategory: 'download', eventAction: 'click', eventLabel: 'latest source codes', nonIteraction: true });">Latest version</a>
<li><a href="https://github.com/deweylab/RSEM/archive/v1.1.17.tar.gz">Version used by RSEM Galaxy Wrapper (RSEM v1.1.17)</a>
<li><a href="https://github.com/deweylab/RSEM/releases">Archive of older versions</a>
<li><a href="https://github.com/deweylab/RSEM">RSEM GitHub
    repository<a/>
</ul>
<p><strong>Documentation</strong></p>
<p><a href="README.html">README</a></p>
<p><strong>Prebuilt RSEM Indices (RSEM v1.1.17) for Galaxy Wrapper</strong></p>
<p>These indices are based on RefSeq containing NM accession numbers only.
That means only curated genes (no experimental, no miRNA, no noncoding).
Only mature RNAs. In addition, 125bp Poly(A) tails are added at the end of each transcript.</p>
<p><a href="http://deweylab.biostat.wisc.edu/rsem/mouse_refseq_NMonly_125bpPolyATail_extractedFromMouseGenome_mm9.tar.gz">Mouse Indices</a>, extracted from mouse genome mm9</p>
<p><a href="http://deweylab.biostat.wisc.edu/rsem/human_refseq_NMonly_125bpPolyATail_extractedFromHumanGenome_hg18.tar.gz">Human Indices</a>, extracted from human genome hg18</p>
<p><strong>Reference annotations and Simulation Data used in the paper</strong></p>
<p><a href="http://deweylab.biostat.wisc.edu/rsem/reference_gtfs.tar.gz">RefSeq and Ensembl annotation GTF files used in the paper</a></p>
<p><a href="http://deweylab.biostat.wisc.edu/rsem/simulation_data.tar.gz">Simulation Data using Refseq set as reference</a></p>
<p><a href="http://deweylab.biostat.wisc.edu/rsem/simulation_data_ensembl.tar.gz">Simulation Data using Ensembl set as reference</a></p>
<p><strong>Google Users and Announce Groups</strong></p>
<table border=0 style="background-color: #fff; padding: 5px;" cellspacing=0>
  <tr><td>
  <img src="groups_logo_sm.gif"
         height=30 width=140 alt="Google Groups">
  </td></tr>
  <tr><td style="padding-left: 5px">
  <b>Subscribe to RSEM Announce</b>
  </td></tr>
  <form action="http://groups.google.com/group/rsem-announce/boxsubscribe">
  <tr><td style="padding-left: 5px;">
  Email: <input type=text name=email>
  <input type=submit name="sub" value="Subscribe">
  </td></tr>
</form>
<tr><td align=right>
  <a href="http://groups.google.com/group/rsem-announce">Visit this group</a>
</td></tr>
</table>
<table border=0 style="background-color: #fff; padding: 5px;" cellspacing=0>
  <tr><td>
  <img src="groups_logo_sm.gif"
         height=30 width=140 alt="Google Groups">
  </td></tr>
  <tr><td style="padding-left: 5px">
  <b>Subscribe to RSEM Users</b>
  </td></tr>
  <form action="http://groups.google.com/group/rsem-users/boxsubscribe">
  <tr><td style="padding-left: 5px;">
  Email: <input type=text name=email>
  <input type=submit name="sub" value="Subscribe">
  </td></tr>
</form>
<tr><td align=right>
  <a href="http://groups.google.com/group/rsem-users">Visit this group</a>
</td></tr>
</table>
<hr>
<p>(last modified on Nov 6, 2015)</p>
</body>
</html>
